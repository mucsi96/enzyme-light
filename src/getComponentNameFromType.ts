import { ElementType } from 'react';

export function getComponentNameFromType(type: ElementType): string {
  if (typeof type === 'function') {
    return (type as any).displayName || type.name || 'null';
  }

  if (typeof type === 'string') {
    return type;
  }
  return 'null';
}
