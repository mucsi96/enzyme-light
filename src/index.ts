export * from './shallow';
export * from './mount';
export * from './ShallowWrapper';
export * from './ReactWrapper';
export * from './waitFor';
export * from './Serializer';
