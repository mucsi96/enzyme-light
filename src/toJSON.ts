import {
  ReactTestInstance,
  ReactTestRendererJSON,
  ReactTestRendererNode,
} from "react-test-renderer";
import { getComponentNameFromType } from "./getComponentNameFromType";

export function toJSON(
  inst: ReactTestInstance | string
): ReactTestRendererNode | null {
  if (typeof inst === "string") {
    return inst;
  }

  const { children, ...props } = inst.props;
  let renderedChildren = null;
  if (inst.children && inst.children.length) {
    for (let i = 0; i < inst.children.length; i++) {
      const renderedChild = toJSON(inst.children[i]);
      if (renderedChild !== null) {
        if (renderedChildren === null) {
          renderedChildren = [renderedChild];
        } else {
          renderedChildren.push(renderedChild);
        }
      }
    }
  }
  const json: ReactTestRendererJSON = {
    type: getComponentNameFromType(inst.type),
    props: props,
    children: renderedChildren,
  };
  Object.defineProperty(json, "$$typeof", {
    value: Symbol.for("react.test.json"),
  });
  return json;
}
