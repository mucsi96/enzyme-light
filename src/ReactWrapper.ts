import { ReactTestInstance, ReactTestRenderer } from 'react-test-renderer';
import { toJSON } from './toJSON';

export class ReactWrapper {
  private rootRenderer: ReactTestRenderer;
  private instances?: ReactTestInstance[];

  constructor(
    rootRenderer: ReactTestRenderer,
    instances?: ReactTestInstance[]
  ) {
    this.rootRenderer = rootRenderer;
    this.instances = instances;
  }

  toJSON() {
    return this.instances?.length
      ? toJSON(this.instances[0])
      : this.rootRenderer.toJSON();
  }

  find<T>(selector: React.ElementType<T>) {
    return new ReactWrapper(
      this.rootRenderer,
      this.rootRenderer.root.findAllByType(selector)
    );
  }

  exists() {
    return (this.instances?.length ?? 0) > 0;
  }
}
