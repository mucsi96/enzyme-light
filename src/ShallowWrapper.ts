import { ReactElement } from 'react';
import { ReactTestInstance } from 'react-test-renderer';
import { toJSON } from './toJSON';

export class ShallowWrapper {
  private renderOutput: ReactElement;
  private instances?: ReactTestInstance[];

  constructor(
    renderOutput: ReactElement,
    instances?: ReactTestInstance[]
  ) {
    this.renderOutput = renderOutput
    this.instances = instances;
  }

  toJSON() {
    return this.instances?.length
      ? toJSON(this.instances[0])
      : this.renderOutput;
  }

  // find<T>(selector: React.ElementType<T>) {
  //   return new ShallowWrapper(
  //     this.renderOutput,
  //     this.rootRenderer.root.findAllByType(selector)
  //   );
  // }

  exists() {
    return (this.instances?.length ?? 0) > 0;
  }
}
