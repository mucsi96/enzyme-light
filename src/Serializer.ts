import { ReactWrapper } from './ReactWrapper';

export class Serializer {
  test(value: ReactWrapper): boolean {
    return value instanceof ReactWrapper;
  }
  
  serialize(val: ReactWrapper, config: any): string {
    const json = val.toJSON();
    const plugin = config.plugins.find((p: any) => p != this && p.test(json));
    return plugin.serialize(json, ...[...arguments].slice(1));
  }
}
