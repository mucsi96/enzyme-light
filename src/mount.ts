import { ReactElement } from 'react';
import { create } from 'react-test-renderer';
import { ReactWrapper } from './ReactWrapper';

export function mount(element: ReactElement): ReactWrapper {
  return new ReactWrapper(create(element));
}
