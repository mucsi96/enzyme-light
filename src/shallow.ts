import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import { ShallowWrapper } from "./ShallowWrapper";

export function shallow(element: ReactElement): ShallowWrapper {
    const shallowRenderer = createRenderer();
    shallowRenderer.render(element);
    return new ShallowWrapper(shallowRenderer.getRenderOutput());
}