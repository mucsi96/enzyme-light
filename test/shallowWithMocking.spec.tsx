import { describe, expect, test, vi } from "vitest";
import { shallow } from "../src";
import { ContactsList } from "./ContactsList";

describe("shallowWithMocking", () => {
  test("renders", async () => {
    const wrapper = shallow(<ContactsList />);
    expect(wrapper).toMatchSnapshot();
  });
});
