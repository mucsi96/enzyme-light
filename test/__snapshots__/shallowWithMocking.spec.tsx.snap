// Vitest Snapshot v1

exports[`shallowWithMocking > renders 1`] = `<ul />`;
